File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int sum = 0, answer = 0;
        unordered_map<int, int> pref;
        pref[sum]++;
        for(int i = 0; i < nums.size(); ++i){
            sum += nums[i];
            if(pref.count(sum - k)){
                answer += pref[sum - k];
            }
            pref[sum]++;
        }
        return answer;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();

        int currSum = 0;
        int answer = 0;
        for (int i = 0; i < nums.length; ++i) {
            currSum += nums[i];
            if (currSum == k) answer += 1;
            if (map.containsKey(currSum - k)) {
                answer += map.get(currSum - k);
            }
            map.put(currSum, map.getOrDefault(currSum, 0) + 1);
        }
        return answer;
    }
}